File: dotnet/Services/LocationService.cs
using System.Collections.Generic;
using System.Linq;
using dungeonbase.Domains;
using dungeonbase.Models;
using dungeonbase.Repositories;

namespace dungeonbase.Services
{
    public class LocationService
    {
        private readonly LocationRepository _locationRepository;

        public LocationService(LocationRepository locationRepository)
        {
            _locationRepository = locationRepository;
        }

        public Location FindById(string id)
        {
            var record = _locationRepository.Get(id);
            var related = GetRelated(record);
            
            return new Location
            {
                Id = record.Id,
                Description = record.Description,
                Name = record.Name,
                Related = related
            };
        }

        private List<LocationRecord> GetRelated(LocationRecord record)
        {
            return record.Related
                .Select(name => _locationRepository.GetByName(name))
                .ToList();
        }

        public List<Location> GetAll()
        {
            return _locationRepository.Get().Select(record =>
            {
                var related = GetRelated(record);
                return new Location
                {
                    Id = record.Id,
                    Description = record.Description,
                    Name = record.Name,
                    Related = related
                };
            }).ToList();
        }
    }
}

File: sprintboot/src/main/java/com/vizientinc/dungeonbase/services/LocationService.java
package com.vizientinc.dungeonbase.services;

import com.vizientinc.dungeonbase.handlers.exceptions.ResourceNotFound;
import com.vizientinc.dungeonbase.models.LocationRecord;
import com.vizientinc.dungeonbase.repositories.ItemRepository;
import com.vizientinc.dungeonbase.repositories.LocationRepository;
import com.vizientinc.dungeonbase.responses.LocationResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import static java.util.stream.Collectors.toList;

@Service
public class LocationService {
    private final LocationRepository locationRepository;
    private final PlayerService playerService;
    private final ItemRepository itemRepository;

    @Autowired
    public LocationService(
        LocationRepository locationRepository,
        PlayerService playerService,
        ItemRepository itemRepository
    ) {
        this.locationRepository = locationRepository;
        this.playerService = playerService;
        this.itemRepository = itemRepository;
    }

    public LocationResponse findById(
        String id
    ) throws Exception {
        LocationRecord locationRecord = locationRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFound(
                "Location",
                id
            ));

        return new LocationResponse(
            locationRecord,
            locationRecord.getRelated()
                .stream()
                .map(locationRepository::findByName)
                .collect(toList()),
            playerService.findByLocation(locationRecord.getId()),
            itemRepository.findAllByLocation(locationRecord.getId())
        );
    }

    public LocationResponse findByPlayerId(String playerId) throws Exception {
        return findById(playerService.findById(playerId)
                            .getLocationId());
    }
}
